internal/auditd/dirreader.go

```diff
@@ -337,37 +340,135 @@ func (o *osFileSystem) Open(filePath string) (io.ReadSeekCloser, error) {
 	return os.Open(filePath)
 }
 
-// readFilePathLines reads all the data from filePath using bufio.ScanLines.
-// It writes each line to the lines chan and returns the number of bytes
-// read from reader.
+// readFilePathLines reads from filePath and writes each line to
+// the lines chan. It then returns the total number of bytes read
+// in terms of lines processed (i.e., if it reads 100 bytes, but
+// only 90 of those bytes were lines ending with a newline, it
+// returns 90). Buffered data is not counted in the byte count.
+//
+// It is similar to os.ReadFile - but avoids reading the entire
+// file's contents into memory at once.
+//
+// An io.EOF error is not returned to the caller.
 func readFilePathLines(ctx context.Context, fsi fileSystem, filePath string, l chan<- string) (int64, error) {
 	f, err := fsi.Open(filePath)
 	if err != nil {
 		return 0, err
 	}
 	defer f.Close()
 
-	return readLines(ctx, f, l)
-}
-
-// readLines reads data from reader until EOF using bufio.ScanLines.
-// It writes each line to the lines chan and returns the number of
-// bytes read from reader.
-func readLines(ctx context.Context, reader io.Reader, lines chan<- string) (int64, error) {
-	scanner := bufio.NewScanner(reader)
+	scanner := bufio.NewScanner(f)
 	var numBytesRead int64
 
 	for scanner.Scan() {
 		line := scanner.Text()
-		numBytesRead += int64(len(line) + 1)
+		numBytesRead += int64(len(line)) + 1
 
 		select {
 		case <-ctx.Done():
 			return numBytesRead, ctx.Err()
-		case lines <- line:
+		case l <- line:
 			// continue.
 		}
 	}
 
 	return numBytesRead, scanner.Err()
 }
+
+// readLines reads from reader and writes each line to the lines chan.
+// It then returns the total number of bytes read in terms of lines
+// processed (i.e., if it reads 100 bytes, but only 90 of those bytes
+// were lines ending with a newline, it returns 90). Buffered data is
+// not counted in the byte count.
+//
+// The supplied backoffSleeper is used to gradually backoff reads that
+// result in an io.EOF. The assumption is that the caller will execute
+// this function again when more data is available to read. Making such
+// a determination is left up to the caller.
+//
+// An io.EOF error is not returned to the caller.
+func readLines(ctx context.Context, reader io.Reader, lines chan<- string, backoff *backoffSleeper) (int64, error) {
```

Comment:
This probably needs a unit test.

Reply:
Are there any specific test scenarios you had in mind? There are three unit tests that directly test this function in `internal/auditd/dirreader_test.go`. There are several indirect tests of this function in that file as well.